### DISCUSSION TITLE
How to use UVtools.AvaloniaControls

### DISCUSSION THREAD

**Discussion by howzew** [sn4k3/UVtools#946] [Q&A] [answered]
I only see one note:
In your app.axaml add:

```
<Application.Styles>
    <StyleInclude Source="avares://UVtools.AvaloniaControls/Controls.axaml"/>
</Application.Styles>
```

I saw the introduction of this control from Avalonia, and it seems to be exactly what I need, because I just want an image control that can zoom, click, and select range, but I don't know how to add this control to Axmal, and I don't know how to get data about clicks, selection areas, etc., can you provide a simple instruction for using it?

**Comment by sn4k3** [comment, marked as answer]
In fact, there are no detailed [documentation](https://github.com/sn4k3/UVtools/tree/master/UVtools.AvaloniaControls), but after add the styles as you did, it works like any other control you may add to axaml, by default it will work like the `Image` Control with zoom.

UVtools heavy use the `AdvancedImageBox`, you can see how I add [here](https://github.com/sn4k3/UVtools/blob/ea2a332edbbe313b2fb42abd4e0b87a2531b68e8/UVtools.UI/MainWindow.axaml#L2118C29-L2126)

After add control in axaml and with Visual Studio you can `Ctrl + Space` inside it propreties line to bring all properties you can set under axaml.
About events and other manipulations you can search [here](https://github.com/sn4k3/UVtools/blob/master/UVtools.UI/MainWindow.LayerPreview.cs) for the keyword: `LayerImageBox`

Try to make it work out of the box, and then to your use, if you have any difficulty you can post here.

**Reply by howzew** [reply]
Thank you very much for your reply, I looked at the code you mentioned, and what confuses me is here:

I can't load images in axaml:

```
<uvtava:AdvancedImageBox 
         Name="LayerImageBox"
         Image="C:\Test.jpg">
</uvtava:AdvancedImageBox>
```

Images also cannot be loaded in behind-code:

```
  public MainWindow()
  {
      InitializeComponent();

      Bitmap img = new Bitmap("C:\Test.jpg");

      LayerImageBox.Image = img;
  }
```

All tips:System.NullReferenceException:“Object reference not set to an instance of an object.”

Is it because the image format is not supported?

**Reply by sn4k3** [reply]
I have tested that usage, it's a bug when you set image in constructor, some UI components are not yet loaded, because they are init after load the theme. 
It's now fixed, you can update the dependency and rebuild.

**Reply by howzew** [reply]
I've updated and retested and I'm having some issues:

```
<Grid Grid.Column="0">
	<uvtava:AdvancedImageBox
		Name="LayerImageBox"
		AutoPan="False"  ShowGrid="True"
		Image="C:\test.jpg">
	</uvtava:AdvancedImageBox>
</Grid>
```

1.This control doesn't seem to have a Stretch="Uniform" property like Image, I'm setting LayerImageBox.ZoomToFit() in behind-code; I don't know if this is the same as Stretch="Uniform", or if it's the right way, but if you set it that way, it seems that there is a problem, when the program starts, the image doesn't fill the whole area (you can see a lot of squares in the background), but when I scroll the mouse wheel casually, the image fills the whole thing<Grid>, and no matter how I scroll the mouse, it can't go back to its original size. I don't understand who controls the size of the initial image and why the Grid can't be populated.

![QQ20241205-125447](https://github.com/user-attachments/assets/1166663f-a771-443e-b02f-c1420dc9405a)
![QQ20241205-125516](https://github.com/user-attachments/assets/207c7b80-3e67-45c3-a3a2-dd01eaa11180)


2.According to the source code, when I right-click, the image should automatically adapt to the size of the control, but when I right-click, nothing happens, am I not turning on a switch?

```
private void LayerImageBoxOnPointerPressed(object? sender, PointerPressedEventArgs e)
    {
        if (e.ClickCount != 2 || (e.KeyModifiers & KeyModifiers.Alt) != 0 || (e.KeyModifiers & KeyModifiers.Shift) != 0) return;
        var pointer = e.GetCurrentPoint(LayerImageBox);
        if (pointer.Properties.IsLeftButtonPressed)
        {
                 .........................
        }

        if (pointer.Properties.IsRightButtonPressed)
        {
            ZoomToFit();
            return; 
        }
        e.Handled = true;
    }
```

3.The function I want to achieve:after drawing the rectangle with the left mouse button, click the right button to clear the rectangle, and I find the following code:

```
private void LayerImageBoxOnPointerMoved(object? sender, PointerEventArgs e)
    {
          .................................

        if (pointer.Properties.IsRightButtonPressed)
        {
            // Right or Alt-Left will remove a pixel
            DrawPixel(false ^ (e.KeyModifiers & KeyModifiers.Alt) != 0, location, e.KeyModifiers);
            return;
        }

        if (pointer.Properties.IsLeftButtonPressed)
        {
            // Left or Alt-Right will add a pixel
            DrawPixel(true ^ (e.KeyModifiers & KeyModifiers.Alt) != 0, location, e.KeyModifiers);
            return;
        }
    }
```

It seems that when I draw a rectangle with the right mouse button, the rectangle should be cleared when I release the mouse, but it turns out that it is not cleared. I don't know how to clear the rectangle.

4.Some properties, such as SelectionMode, can only be set in behind-code and cannot be set in axaml.

Sorry I have so many questions.

**Reply by sn4k3** [reply]
> 1.This control doesn't seem to have a Stretch="Uniform" property like Image,

By default the control render the image in Normal mode (1:1), which respects the original image size without any manipulation.
To change modes, define `SizeMode` property to any of [this enum](https://github.com/sn4k3/UVtools/blob/master/UVtools.AvaloniaControls/AdvancedImageBox.cs#L363)
With a Stretch mode you will lose zoom capabilities because it always fit your view, so my recommendation is to subscribe the Image property and when it change call the `ZoomToFit`.

> 2.According to the source code, when I right-click, the image should automatically adapt to the size of the control, but when I right-click, nothing happens, am I not turning on a switch?

You are seeing the source code of my main app, not the AdvancedImage library. The library has no such built-in shortcut.
You must implement on your app your own shortcuts like I did with UVtools.

3. It must be implemented in your code too, you can draw directly on bitmap and track mouse coords.
However if you are talking about the selection tool you can clear it via: `LayerImage.SelectNone()`

> 4.Some properties, such as SelectionMode, can only be set in behind-code and cannot be set in axaml.

They must work, they are StyledProperties, eg: `SelectionMode="Zoom"` in axaml works for me.

**Reply by howzew** [reply]
Thank you very much for your reply. Regarding the first point, I still don't understand:

1. Why is the image displayed in the smaller middle area after calling LayerImageBox.ZoomToFit();, and why is there a lot of blank space around it? I didn't set the width and height for AdvancedImageBox.After calling ZoomToFit(), it looks like picture 1, but I think it should look like picture 2, because I think it should fit the size of the Grid.

2. If I want the image to automatically adapt to the size of the Grid after loading and keep the zoom function, how should I do?Just like the Image control, it is as simple as using Uniform but it does not affect whether it can be Zoom in/out. I wonder if AdvancedImageBox has the same simple setting.

**Reply by sn4k3** [reply]
1. Where do you call ` LayerImageBox.ZoomToFit();`? If in constructor it's normal that it won't work, everything in avalonia and related to bounds and auto sizes never work in constructor as layout is not yet created/rendered into UI, this way auto sizes can't be calculated in that stage. You need to override the `OnLoaded` of your own control and call there the center function. In future I can include a property to auto zoom to fit when loading a new image.

2. Looks like you new to avalonia? Every avalonia property can be subscribed for changes, as so you must subscribe `Image` property, and call the function. Or you can simply register the event `PropertyChanged` and look for `Image`, eg:

```C#
LayerImageBox.PropertyChanged += (sender, e) =>
{
    if (e.PropertyName == nameof(AdvancedImageBox.Image))
    {
        LayerImageBox.ZoomToFit();
    }
};
```

Keep in mind if you set image on xmal or constructor you need to call `LayerImageBox.ZoomToFit();` on your `OnLoaded` override, because of 1.

**Reply by howzew** [reply]
You are right, I first called this method in the constructor.
Thank you very much for your help. I did not subscribe to the `Image` property, but directly called `LayerImageBox.ZoomToFit();` in the `override OnLoaded` method to successfully achieve this effect.

```
  protected override void OnLoaded(RoutedEventArgs e)
  {
      LayerImageBox.ZoomToFit();
  }
```

In addition, thank you for your efforts in this control. This is really a very convenient and efficient control. It solved my big problem. Thank you again!

**Reply by sn4k3** [reply]
I have made some improvements and released an updated version:

```
- Pan with keys now only works when no modifier is pressed
- Change property name from `AllowZoom` to `ZoomWithMouseWheel`
- Fix `ZoomLevelToFit` to never return values lower than 0, it will return `100` in that case
- Fix `ConstrainZoomOutToFitLevel` to constrain when loading an image
- Fix ViewPort vertical scrollbar calculation and constrain to positive values
- Change `ConstrainZoomOutToFitLevel` default from `true` to `false`
- Improve ViewPort update when loading images in constructor
- Add `AutoZoomToFit` property
- Add `ZoomInKeyGestures` property
- Add `ZoomOutKeyGestures` property
- Add `ZoomTo100KeyGestures` property
- Add `ZoomToFitKeyGestures` property
- Add `AutoZoomToFit` property
- Add `AutoZoomToFit` property
```

new use example:

```xml
AutoZoomToFit="True"
ConstrainZoomOutToFitLevel="True"
ZoomInKeyGestures="Ctrl+OemPlus, Ctrl+Add"
ZoomOutKeyGestures="Ctrl+OemMinus, Ctrl+Subtract"
ZoomTo100KeyGestures="Ctrl+D0, Ctrl+NumPad0"
ZoomToFitKeyGestures="Ctrl+Alt+D0, Ctrl+Alt+NumPad0"
```

**Reply by howzew** [reply]
It's awesome and it's so much easier to use!
